### PR TITLE
Don't proceed into setup.py too far if Python version is unsupported

### PR DESCRIPTION
**Pull request by xuhdev**
This prevents confusing errors when the interpreter encounters some
syntax errors in the middle.

### CODE REVIEW

**Comment by xuhdev** [comment]
Please let me know if we also want to cap the maximum version here. For example, this confusion https://github.com/pytorch/pytorch/issues/42832#issuecomment-671673440 seems to be caused by Python 3.9, greater than our maximum supported version: https://github.com/pytorch/pytorch/issues/42832#issuecomment-671682721

The downside is that we might have to also allow people to experiment newer version of Python by giving them some "backdoor" to go through.

File: setup.py [hunk 1]
```diff
@@ -171,6 +171,14 @@
     print("32-bit Windows Python runtime is not supported. Please switch to 64-bit Python.")
     sys.exit(-1)
 
+import platform
+python_min_version = (3, 6, 1)
+python_min_version_str = '.'.join((str(num) for num in python_min_version))
+if sys.version_info < python_min_version:
+    print("You are using Python {}. Python >={} is required.".format(platform.python_version(),
+                                                                     python_min_version_str))
+    sys.exit(-1)
```
Comment (nbcsm) [inline setup.py:180]: also check the max version?

**Review by nbcsm — APPROVED** [review]

File: setup.py [hunk 2]
```diff
@@ -171,6 +171,17 @@
     print("32-bit Windows Python runtime is not supported. Please switch to 64-bit Python.")
     sys.exit(-1)
 
+import platform
+python_min_version = (3, 6, 1)
+python_min_version_str = '.'.join((str(num) for num in python_min_version))
+python_max_version = (3, 9, 0)
+python_max_version_str = '.'.join((str(num) for num in python_max_version))
```
Comment (ezyang) [inline setup.py:178]: I'm surprised there isn't a more compact way to write this code

Comment (xuhdev) [reply setup.py:178]: me, too (not I'm aware of)

Comment (nbcsm) [reply setup.py:178]: https://pypi.org/project/semver/ may make the code simple, but need to introduce a new dependency...

**Comment by ezyang** [comment]
The max version check seems unnecessary and going to force us to keep modifying this bound, even if we are actually compatible with a new version of Python?

**Comment by xuhdev** [comment]
Yeah, that's the downside of having a maximum version imposed. I think it really depends on what we would like to achieve.

Pros:

- Avoid users experience unsupported Python version.
- Users won't face strange newer-Python version bugs before we have tested.

Cons:

- Prevent users from experimenting.

Of course, we can add a switch, but I think that would be too complicated for this purpose.

Please let me know which way the PyTorch team would like to go; I'm fine with either.

File: setup.py (hunk 2 above)
Comment (malfet) [reply setup.py:178]: How about the following?
```suggestion
python_max_version_str = '.'.join(map(str, python_max_version))
```

File: setup.py [hunk 3]
```diff
@@ -171,6 +171,17 @@
...
+if sys.version_info < python_min_version or sys.version_info >= python_max_version:
+    print("You are using Python {}. Python >={},<{} is required.".format(platform.python_version(),
+                                                                         python_min_version_str,
+                                                                         python_max_version_str))
+    sys.exit(-1)
```
Comment (malfet) [inline setup.py:183]: Hmm, why not use some python3 features like f-strings here?
```suggestion
def tuple2str(t, sep='.'):
   return sep.join(map(str, t))
python_min_version = (3, 6, 1)
python_max_version = (3, 9, 0)
if sys.version_info < python_min_version or sys.version_info >= python_max_version:
    print(f"You are using Python {tuple2str(sys.version_info[:3]}. Python >={tuple2str(python_min_version)},<{tuple2str(python_max_version)} is required.")
    sys.exit(-1)
```

Comment (xuhdev) [reply setup.py:183]: You would receive syntax error if Python is < 3.6, and that's exactly the situation where we are trying to have a reasonable error message.

**Comment by xuhdev** [comment]
So... would we proceed?

**Comment by ezyang** [comment]
I'm going to pull out my veto power, and say, purely on the perspective of not wanting to have to update hardcoded Python version every time a new Python version comes out, let's get rid of the upper bound.

File: setup.py [hunk 4]
```diff
@@ -883,26 +894,23 @@ def print_box(msg):
...
             'Topic :: Scientific/Engineering',
             'Topic :: Scientific/Engineering :: Mathematics',
             'Topic :: Scientific/Engineering :: Artificial Intelligence',
             'Topic :: Software Development',
             'Topic :: Software Development :: Libraries',
             'Topic :: Software Development :: Libraries :: Python Modules',
-        ],
+            'Programming Language :: C++',
+            'Programming Language :: Python :: 3',
+        ] + ['Programming Language :: Python :: 3.{}' for i in range(python_min_version[1], python_max_version[1])],
```
Comment (xuhdev) [inline setup.py:913]: @ezyang Do you think this is a reasonable solution? You've got to change the classifier any way.

**Pull request closed without merging**